File: mobile/android/src/main/kotlin/com/otravelz/android/ui/screens/MapViewModel.kt
package com.otravelz.android.ui.screens

import android.annotation.SuppressLint
import android.content.Context
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.google.android.gms.location.LocationServices
import com.google.android.gms.location.Priority
import com.otravelz.android.BuildConfig
import com.otravelz.android.data.network.ApiClient
import com.otravelz.android.domain.model.*
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

data class MapUiState(
    val productState: MapProductState = MapProductState.Loading,
    val layers: MapLayersState = MapLayersState(),
    val destinations: List<DiscoverPlace> = emptyList(),
    val verifiedStops: List<TransitStopMarker> = emptyList(),
    val candidateStops: List<TransitStopMarker> = emptyList(),
    val civicServices: List<CivicServiceMarker> = emptyList(),
    val selectedRoute: TransitRouteGeometry? = null,
    val selectedEntity: SelectedMapEntity = SelectedMapEntity.None,
    val locationStatus: LocationStatus = LocationStatus.Unknown,
    val searchQuery: String = "",
    val selectedDistrict: String? = null,
    val isListAlternativeVisible: Boolean = false,
    val cameraTarget: GeoCoordinate = GeoCoordinate.ODISHA_CENTER,
    val cameraZoom: Float = 7.2f
) {
    /**
     * Filtered destinations according to search query, selected district, and layers.
     */
    val visibleDestinations: List<DiscoverPlace>
        get() {
            if (!layers.showDestinations) return emptyList()
            var result = destinations
            if (selectedDistrict != null) {
                result = result.filter { it.normalizedDistrict.equals(selectedDistrict, ignoreCase = true) }
            }
            if (searchQuery.isNotBlank()) {
                val q = searchQuery.trim().lowercase()
                result = result.filter { place ->
                    place.name.lowercase().contains(q) ||
                    (place.odiaName?.lowercase()?.contains(q) == true) ||
                    place.category.lowercase().contains(q) ||
                    (place.district?.lowercase()?.contains(q) == true)
                }
            }
            return result
        }

    val visibleStops: List<TransitStopMarker>
        get() {
            val list = mutableListOf<TransitStopMarker>()
            if (layers.showVerifiedStops) {
                list.addAll(verifiedStops.filter { it.canRenderMarker })
            }
            if (layers.showCandidateStops) {
                list.addAll(candidateStops.filter { it.canRenderCandidateMarker })
            }
            return list
        }

    val visibleServices: List<CivicServiceMarker>
        get() = if (layers.showEssentials) civicServices else emptyList()
}

class MapViewModel : ViewModel() {

    private val apiService = ApiClient.createService()
    private val _uiState = MutableStateFlow(MapUiState())
    val uiState: StateFlow<MapUiState> = _uiState.asStateFlow()

    init {
        checkProviderAndLoadData()
    }

    fun checkProviderAndLoadData() {
        val key = BuildConfig.MAPS_API_KEY
        if (key.isBlank()) {
            _uiState.update {
                it.copy(
                    productState = MapProductState.ProviderUnavailable(
                        "Google Maps API key is unconfigured. Set MAPS_API_KEY in local.properties or environment."
                    )
                )
            }
        }
        loadDestinations()
    }

    fun loadDestinations() {
        viewModelScope.launch {
            try {
                val dtos = apiService.getPlaces(limit = 250)
                val places = dtos.map { it.toDiscoverPlace() }
                    .filter { it.isEligibleLeisure && it.hasCoordinates }
                _uiState.update { current ->
                    current.copy(
                        destinations = places,
                        productState = if (BuildConfig.MAPS_API_KEY.isBlank()) {
                            current.productState
                        } else {
                            MapProductState.Ready
                        }
                    )
                }
            } catch (e: Exception) {
                _uiState.update { current ->
                    if (current.destinations.isEmpty()) {
                        current.copy(productState = MapProductState.DataUnavailable(e.message ?: "Failed to load destinations"))
                    } else {
                        current
                    }
                }
            }
        }
    }

    fun toggleDestinationsLayer() {
        _uiState.update { it.copy(layers = it.layers.toggleDestinations()) }
    }

    fun toggleEssentialsLayer() {
        val nextLayers = _uiState.value.layers.toggleEssentials()
        _uiState.update { it.copy(layers = nextLayers) }
        if (nextLayers.showEssentials && _uiState.value.civicServices.isEmpty()) {
            loadNearbyServices(_uiState.value.cameraTarget.lat, _uiState.value.cameraTarget.lon)
        }
    }

    fun toggleVerifiedStopsLayer() {
        val next = !_uiState.value.layers.showVerifiedStops
        _uiState.update { it.copy(layers = it.layers.copy(showVerifiedStops = next)) }
        if (next && _uiState.value.verifiedStops.isEmpty()) {
            loadNearbyStops(_uiState.value.cameraTarget.lat, _uiState.value.cameraTarget.lon)
        }
    }

    fun toggleCandidateStopsLayer() {
        val next = !_uiState.value.layers.showCandidateStops
        _uiState.update { it.copy(layers = it.layers.copy(showCandidateStops = next)) }
    }

    fun selectEntity(entity: SelectedMapEntity) {
        _uiState.update { current ->
            when (entity) {
                is SelectedMapEntity.Destination -> {
                    val coord = GeoCoordinate.fromOrNull(entity.place.lat, entity.place.lon)
                    current.copy(
                        selectedEntity = entity,
                        cameraTarget = coord ?: current.cameraTarget,
                        cameraZoom = if (coord != null) 14.0f else current.cameraZoom
                    )
                }
                is SelectedMapEntity.TransitStop -> {
                    current.copy(
                        selectedEntity = entity,
                        cameraTarget = entity.stop.coordinate ?: current.cameraTarget,
                        cameraZoom = if (entity.stop.coordinate != null) 15.0f else current.cameraZoom
                    )
                }
                is SelectedMapEntity.CivicService -> {
                    current.copy(
                        selectedEntity = entity,
                        cameraTarget = entity.service.coordinate,
                        cameraZoom = 15.0f
                    )
                }
                is SelectedMapEntity.TransitRoute -> {
                    current.copy(
                        selectedEntity = entity,
                        selectedRoute = entity.route
                    )
                }
                SelectedMapEntity.None -> current.copy(selectedEntity = entity)
            }
        }
    }

    fun clearSelection() {
        _uiState.update { it.copy(selectedEntity = SelectedMapEntity.None) }
    }

    fun onSearchQueryChanged(query: String) {
        _uiState.update { it.copy(searchQuery = query) }
    }

    fun onDistrictSelected(district: String?) {
        _uiState.update { it.copy(selectedDistrict = district) }
    }

    fun toggleListAlternative() {
        _uiState.update { it.copy(isListAlternativeVisible = !it.isListAlternativeVisible) }
    }

    fun loadNearbyServices(lat: Double, lon: Double) {
        if (!GeoCoordinate.isValid(lat, lon)) return
        viewModelScope.launch {
            try {
                val response = apiService.getNearbyServices(lat = lat, lon = lon, radiusKm = 25.0)
                val services = response.services.mapNotNull { CivicServiceMarker.fromDto(it) }
                _uiState.update { it.copy(civicServices = services) }
            } catch (_: Exception) {
                // Calm degradation; empty services list
            }
        }
    }

    fun loadNearbyStops(lat: Double, lon: Double) {
        if (!GeoCoordinate.isValid(lat, lon)) return
        viewModelScope.launch {
            try {
                val list = apiService.getNearbyStops(lat = lat, lon = lon, radiusMeters = 5000)
                val stops = list.map { TransitStopMarker.fromDto(it) }
                val verified = stops.filter { it.tier.isVerifiedPhysicalPole }
                val candidates = stops.filter { it.tier.isCandidate }
                _uiState.update { it.copy(verifiedStops = verified, candidateStops = candidates) }
            } catch (_: Exception) {
                // Calm degradation; empty stops list
            }
        }
    }

    fun onLocationPermissionDenied() {
        _uiState.update {
            it.copy(
                locationStatus = LocationStatus.PermissionDenied,
                layers = it.layers.copy(showUserLocation = false)
            )
        }
    }

    @SuppressLint("MissingPermission")
    fun requestUserLocation(context: Context) {
        try {
            val fusedClient = LocationServices.getFusedLocationProviderClient(context)
            fusedClient.getCurrentLocation(Priority.PRIORITY_BALANCED_POWER_ACCURACY, null)
                .addOnSuccessListener { loc ->
                    if (loc != null && GeoCoordinate.isValid(loc.latitude, loc.longitude)) {
                        val live = LocationStatus.Live(loc.latitude, loc.longitude, loc.accuracy)
                        _uiState.update {
                            it.copy(
                                locationStatus = live,
                                layers = it.layers.copy(showUserLocation = true),
                                cameraTarget = GeoCoordinate(loc.latitude, loc.longitude),
                                cameraZoom = 13.5f
                            )
                        }
                    } else {
                        _uiState.update { it.copy(locationStatus = LocationStatus.LocationUnavailable) }
                    }
                }
                .addOnFailureListener {
                    _uiState.update { it.copy(locationStatus = LocationStatus.LocationUnavailable) }
                }
        } catch (_: SecurityException) {
            _uiState.update { it.copy(locationStatus = LocationStatus.PermissionDenied) }
        }
    }
}

File: mobile/ios/OTravelz/Features/MapViewModel.swift
import Foundation
import CoreLocation
import MapKit
import SwiftUI

@Observable
@MainActor
final class MapViewModel: NSObject, CLLocationManagerDelegate {
    var productState: MapProductState = .loading
    var layers: MapLayersState = MapLayersState()
    var destinations: [PlaceCardModel] = []
    var verifiedStops: [TransitStopMarker] = []
    var candidateStops: [TransitStopMarker] = []
    var civicServices: [CivicServiceMarker] = []
    var selectedEntity: SelectedMapEntity = .none
    var locationStatus: LocationStatus = .unknown
    var searchQuery: String = ""
    var selectedDistrict: String? = nil
    var isListAlternativeVisible: Bool = false
    var cameraPosition: MapCameraPosition = .region(
        MKCoordinateRegion(
            center: CLLocationCoordinate2D(latitude: 20.27, longitude: 84.85),
            span: MKCoordinateSpan(latitudeDelta: 4.5, longitudeDelta: 4.5)
        )
    )

    private let locationManager = CLLocationManager()

    override init() {
        super.init()
        locationManager.delegate = self
        locationManager.desiredAccuracy = kCLLocationAccuracyHundredMeters
    }

    var visibleDestinations: [PlaceCardModel] {
        guard layers.showDestinations else { return [] }
        var result = destinations
        if let district = selectedDistrict {
            result = result.filter { $0.district?.caseInsensitiveCompare(district) == .orderedSame }
        }
        if !searchQuery.trimmingCharacters(in: .whitespaces).isEmpty {
            let q = searchQuery.lowercased().trimmingCharacters(in: .whitespaces)
            result = result.filter { place in
                place.name.lowercased().contains(q) ||
                (place.odiaName?.lowercased().contains(q) == true) ||
                place.category.lowercased().contains(q) ||
                (place.district?.lowercased().contains(q) == true)
            }
        }
        return result
    }

    var visibleStops: [TransitStopMarker] {
        var list: [TransitStopMarker] = []
        if layers.showVerifiedStops {
            list.append(contentsOf: verifiedStops.filter { $0.canRenderMarker })
        }
        if layers.showCandidateStops {
            list.append(contentsOf: candidateStops.filter { $0.canRenderCandidateMarker })
        }
        return list
    }

    var visibleServices: [CivicServiceMarker] {
        guard layers.showEssentials else { return [] }
        return civicServices
    }

    func loadDestinations() async {
        productState = .loading
        do {
            let dtos = try await APIClient.shared.fetchPlaces()
            let models = dtos
                .compactMap { PlaceDomainMapper.toCardModel(dto: $0) }
                .filter { $0.isEligibleLeisure && $0.coordinate != nil }
            self.destinations = models
            self.productState = .ready
        } catch {
            if destinations.isEmpty {
                self.productState = .dataUnavailable(error.localizedDescription)
            } else {
                self.productState = .ready
            }
        }
    }

    func toggleDestinations() {
        layers.toggleDestinations()
    }

    func toggleEssentials() {
        layers.toggleEssentials()
    }

    func toggleVerifiedStops() {
        layers.showVerifiedStops.toggle()
    }

    func toggleCandidateStops() {
        layers.showCandidateStops.toggle()
    }

    func toggleListAlternative() {
        isListAlternativeVisible.toggle()
    }

    func selectDestination(_ place: PlaceCardModel) {
        selectedEntity = .destination(place)
        if let coord = place.coordinate {
            cameraPosition = .region(
                MKCoordinateRegion(
                    center: CLLocationCoordinate2D(latitude: coord.latitude, longitude: coord.longitude),
                    span: MKCoordinateSpan(latitudeDelta: 0.05, longitudeDelta: 0.05)
                )
            )
        }
    }

    func selectStop(_ stop: TransitStopMarker) {
        selectedEntity = .transitStop(stop)
        if let coord = stop.coordinate {
            cameraPosition = .region(
                MKCoordinateRegion(
                    center: coord.clCoordinate,
                    span: MKCoordinateSpan(latitudeDelta: 0.02, longitudeDelta: 0.02)
                )
            )
        }
    }

    func selectService(_ service: CivicServiceMarker) {
        selectedEntity = .civicService(service)
        cameraPosition = .region(
            MKCoordinateRegion(
                center: service.coordinate.clCoordinate,
                span: MKCoordinateSpan(latitudeDelta: 0.02, longitudeDelta: 0.02)
            )
        )
    }

    func clearSelection() {
        selectedEntity = .none
    }

    func requestLocation() {
        switch locationManager.authorizationStatus {
        case .notDetermined:
            locationManager.requestWhenInUseAuthorization()
        case .authorizedWhenInUse, .authorizedAlways:
            locationManager.requestLocation()
        case .denied, .restricted:
            locationStatus = .permissionDenied
        @unknown default:
            locationStatus = .unknown
        }
    }

    // MARK: - CLLocationManagerDelegate

    nonisolated func locationManagerDidChangeAuthorization(_ manager: CLLocationManager) {
        Task { @MainActor in
            switch manager.authorizationStatus {
            case .authorizedWhenInUse, .authorizedAlways:
                manager.requestLocation()
            case .denied, .restricted:
                self.locationStatus = .permissionDenied
            default:
                break
            }
        }
    }

    nonisolated func locationManager(_ manager: CLLocationManager, didUpdateLocations locations: [CLLocation]) {
        Task { @MainActor in
            guard let loc = locations.last else { return }
            if GeoCoordinate.isValid(latitude: loc.coordinate.latitude, longitude: loc.coordinate.longitude) {
                self.locationStatus = .live(
                    latitude: loc.coordinate.latitude,
                    longitude: loc.coordinate.longitude,
                    accuracyMeters: loc.horizontalAccuracy
                )
                self.layers.showUserLocation = true
                self.cameraPosition = .region(
                    MKCoordinateRegion(
                        center: loc.coordinate,
                        span: MKCoordinateSpan(latitudeDelta: 0.05, longitudeDelta: 0.05)
                    )
                )
            } else {
                self.locationStatus = .locationUnavailable
            }
        }
    }

    nonisolated func locationManager(_ manager: CLLocationManager, didFailWithError error: Error) {
        Task { @MainActor in
            self.locationStatus = .locationUnavailable
        }
    }
}
